### ISSUE TITLE
Handling decimal numbers is inconvenient

### ISSUE THREAD

**Issue by golszewski86** [piotrkot/oojson#38]
Consider the following code:

```java
final JsonObj json = new JsonObj(
    new StringReader(
        String.format(
            "{ \"old\": 2.50, \"new\": 2.51 }"
        )
    )
);
System.out.println(json.get("old").getClass());
System.out.println(json.get("new").getClass());
```

The result is:

```
class java.math.BigDecimal
class java.lang.Double
```

This means that even if I know that field contains a decimal value, I cannot be sure if it will be Double or BigDecimal. Because of that, handling decimal values is inconvenient because requires either converting all decimal numbers to Double (and possible lost of precision), or using ugly `instanceof` operator to test for value type.

**Comment by piotrkot** [comment]
In the example, you have a string which represents a JSON, and probably based on some documentation you know that both "old" and "new" fields have numerical values.

Well, than both values extend `java.lang.Number`. If you know you are safe you can take double from the Number with `doubleValue()` which indeed may involve rounding. Otherwise just pass the Number argument around.

**Comment by golszewski86** [comment]
I don't need Number. I need a BigDecimal.

**Comment by piotrkot** [comment]
OK. Can't you call `json.<BigDecimal>get("new")` and `json.<BigDecimal>get("old")`? Both values will be of BigDecimal type.

**Comment by golszewski86** [comment]
No they won't. Casting object of type Double to type BigDecimal won't make it BigDecimal; only ClassCastException will be raised. Try it yourself.

**Comment by piotrkot** [comment]
```
    @Test
    public void shouldCastingFail() throws Exception {
        final JsonObj json = new JsonObj(
            new StringReader(
                "{ \"old\": 2.50, \"new\": 2.51 }"
            )
        );
        System.out.println(json.<BigDecimal>get("new"));
        System.out.println(json.<BigDecimal>get("old"));
    }
```
This one works fine.

**Comment by golszewski86** [comment]
And that's exactly the problem.

If I know that particular JSON field will be of type String, I can safely call `json.<String>get("someString")`.
If I know that particular JSON field will be of type Boolean, I can safely call `json.<Boolean>get("someBool")`.

But with decimal number, I can never know whether JSON field containing decimal number will be of type Double or BigDecimal.

The same problem applies to integer values. The type can be Integer, Long or BigInteger depending on how big the number is. And whereas there are methods `Number#intValue` and `Number#longValue`, there is no way to convert Number to BigInteger hence the only safe way to deal with integer numbers is to obtain object via `json.<Number>get("someNum")` and then using the

**Comment by golszewski86** [comment]
Yes, this one works fine because Java compiler is smart enough to remove casting to BigDecimal when it realizes that you don't actually need BigDecimal type. And then it calls println(Object) method.

Try this one:
```
System.out.println(json.<BigDecimal>get("new").scale());
System.out.println(json.<BigDecimal>get("old").scale());
```

The result is:

```
2
Exception in thread "main" java.lang.ClassCastException: java.lang.Double cannot be cast to java.math.BigDecimal
```
